lib/groups.js

```diff
@@ -191,38 +332,148 @@ module.exports.createGroup = function (groupName, assessmentId, userId, authnUse
-  sqldb.call('group_leave', params, function (err) {
-    if (ERR(err, callback)) return;
-    callback();
+/**
+ * @param {string} assessmentId
+ * @param {string} userId
+ * @param {string} authnUserId
+ */
+module.exports.leaveGroup = async function (assessmentId, userId, authnUserId) {
+  await sqldb.runInTransactionAsync(async () => {
+    const groupId = await module.exports.getGroupId(assessmentId, userId);
+    const groupConfig = await module.exports.getGroupConfig(assessmentId);
+
+    if (groupConfig.has_roles) {
+      const groupInfo = await module.exports.getGroupInfo(groupId, groupConfig);
+
+      // Reassign roles if there is more than 1 user
+      const currentSize = groupInfo.groupMembers.length;
+      if (currentSize > 1) {
+        // Get leaving user's required roles
+        const requiredRoles = groupInfo.rolesInfo.groupRoles.filter((role) => role.minimum > 0);
+        const groupRoleAssignments = Object.values(groupInfo.rolesInfo.roleAssignments).flat();
+        const requiredGroupRoleIds = groupRoleAssignments
+          .filter(({ user_id }) => user_id === userId.toString())
+          .map(({ group_role_id }) => group_role_id)
+          .filter((roleId) => requiredRoles.some((role) => role.id === roleId));
+
+        if (requiredGroupRoleIds.length > 0) {
+          // Get group user to group role assignments, excluding the leaving user
+          const groupRoleAssignments = Object.values(groupInfo.rolesInfo.roleAssignments)
+            .flat()
+            .filter(({ user_id }) => user_id !== userId.toString());
+
+          // Replace non-required group role assignments with required roles from leaving user
+          const groupRoleAssignmentUpdates = groupRoleAssignments.map(
+            ({ user_id, group_role_id }) => {
+              const notRequired = !requiredRoles.some((role) => role.id === group_role_id);
+              if (notRequired && requiredGroupRoleIds.length > 0) {
+                return {
+                  user_id,
+                  group_role_id: requiredGroupRoleIds.pop(),
+                  group_id: groupId,
+                };
+              } else {
+                return { user_id, group_role_id, group_id: groupId };
+              }
+            }
+          );
+
+          // Assign remaining required roles to another group user
+          const assigneeUser = groupInfo.groupMembers.find((m) => m.user_id !== userId.toString());
+          for (const groupRoleId of requiredGroupRoleIds) {
+            groupRoleAssignmentUpdates.push({
+              user_id: assigneeUser.user_id,
+              group_role_id: groupRoleId,
+              group_id: groupId,
+            });
+          }
+
+          await sqldb.queryAsync(sql.clear_group_roles, { group_id: groupId });
+          await sqldb.queryAsync(sql.reassign_group_roles_after_leave, {
+            role_assignments: JSON.stringify(groupRoleAssignmentUpdates),
+          });
```

Comment:
You should be able to combine these into a single query using CTEs; you can probably even combine it with `delete_not_required_roles` below. While not strictly necessary, we generally like to limit the number of database round-trips we have to do.